File: server/services/activityService.js
const db = require("../db/jsonStore");

function log(userId, action, entityType, entityId, before = null, after = null) {
  return db.table("activity_logs").insert({
    user_id: userId,
    action,
    entity_type: entityType,
    entity_id: entityId,
    before_state: before,
    after_state: after,
    created_at: new Date().toISOString(),
  });
}

function listRecent(limit = 25) {
  return db
    .table("activity_logs")
    .all()
    .sort((a, b) => new Date(b.created_at) - new Date(a.created_at))
    .slice(0, limit)
    .map((entry) => {
      const user = db.table("users").find(entry.user_id);
      return Object.assign({}, entry, { username: user ? user.username : "system" });
    });
}

module.exports = { log, listRecent };

File: backend/task_management/services/activity_service.py
from ..extensions import db
from ..models import ActivityLog


def log(user_id, action, entity_type, entity_id=None, before=None, after=None):
    entry = ActivityLog(
        user_id=user_id,
        action=action,
        entity_type=entity_type,
        entity_id=entity_id,
        before_state=before,
        after_state=after,
    )
    db.session.add(entry)
    db.session.commit()
    return entry


def list_recent(limit=25):
    entries = (
        ActivityLog.query.order_by(ActivityLog.created_at.desc()).limit(limit).all()
    )
    return [entry.to_dict() for entry in entries]
